### PR TITLE
SALTO-4404: Salesforce: Replace `Task`/`Event` custom fields with reference to the corresponding `Activity` fields

### PR DESCRIPTION
**Pull request by Nurdok**
This PR modifies `Task`/`Event` custom fields, which are uneditable by themselves and controlled by the `Activity` object:
- Added a filter to remove custom field information and replace it with a reference to the corresponding `Activity` custom field
- The same filter drops custom field changes on deploy since they will be applied by the `Activity` custom fields
- Changed the CV to allow adding / removing of custom fields if they have a corresponding `Activity` field which is also added / removed.

#### Before

```hcl
salesforce.Text Test__c {
  apiName = "Event.Test__c"
  length = 18
  label = "Test"
  createable = true
  updateable = true
  unique = false
  caseSensitive = false
  queryable = true
}
```

#### After

```hcl
salesforce.Text Test__c {
  apiName = "Event.Test__c"
  activityField = salesforce.Activity.field.Test__c
}
```

---

_Additional context for reviewer_

---
_Release Notes_: 
TODO

---
_User Notifications_: 
TODO

### CODE REVIEW

**Review by tamtamirr — COMMENTED** [review]
Added some comments :)

File: packages/salesforce-adapter/src/fetch_profile/fetch_profile.ts [hunk 1]
```diff
@@ -40,6 +40,7 @@ const optionalFeaturesDefaultValues: OptionalFeaturesDefaultValues = {
   metaTypes: false,
   cpqRulesAndConditionsRefs: true,
   flowCoordinates: false,
+  // TODO: add
```
[the remark this answers is not in the document]
Comment (Nurdok) [reply packages/salesforce-adapter/src/fetch_profile/fetch_profile.ts]: Done

**Review by tamtamirr — COMMENTED** [review]
Looks good! Added some comments

File: packages/salesforce-adapter/src/filters/utils.ts [hunk 2]
```diff
@@ -945,3 +948,8 @@ export const toCustomProperties = async (
     ..._.pickBy(element.annotations, (_val, name) => isAllowed(name)),
   }
 }
+
+export const isCustomField = (field: Field): boolean => field.name.endsWith(SALESFORCE_CUSTOM_SUFFIX)
+
+export const isFieldOfTaskOrEvent = ({ parent }: Field): boolean =>
+  isCustomObjectSync(parent) && [TASK_CUSTOM_OBJECT, EVENT_CUSTOM_OBJECT].includes(apiNameSync(parent) ?? '')
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/utils.ts:955]: Please add UTs 🙏

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        'Error occurred when attempting to remodel CustomFields of Task and Event to reference their respective Activity fields.',
+      filterName: 'taskAndEventCustomFields',
+      config,
+      fetchFilterFunc: async (elements: Element[]) => {
+        const elementsSource = buildElementsSourceForFetch(elements, config)
+        const elementSourceByApiName = await awu(await elementsSource.getAll())
+          .filter(isDefined)
+          .filter(isCustomObjectSync)
+          .keyBy(customObject => apiNameSync(customObject) ?? '')
+        const activity = elementSourceByApiName[ACTIVITY_CUSTOM_OBJECT]
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:45]: Can be simplified
```suggestion
        const activity = await awu(await elementsSource.getAll())
          .filter(isDefined)
          .filter(isCustomObjectSync)
          .find(customObject => apiNameSync(customObject) === ACTIVITY_CUSTOM_OBJECT)
```
We can also avoid invoking `getAll` here by using `get` instead:
```
const activity = await elementsSource.get(new ElemID(SALESFORCE, ACTIVITY_CUSTOM_OBJECT))
if (!isCustomObjectSync(activity)) {
  return
}
```

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        if (activity === undefined) {
+          return
+        }
+
+        const elementsByApiName = _.keyBy(elements.filter(isCustomObjectSync), elem => apiNameSync(elem) ?? '')
+
+        Object.entries(activity.fields)
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:52]: We should work with the Fields only and not their names. Will discuss that F2F

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    }),
+    preDeploy: async (changes: Change[]): Promise<void> => {
+      changesToRestore = changes
+        .filter(isFieldChange)
+        .filter(change => isFieldOfTaskOrEvent(getChangeData(change)))
+        .filter(change => isCustomField(getChangeData(change)))
+
+      changesToRestore.forEach(change => {
+        changes.splice(changes.indexOf(change), 1)
+      })
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:81]:
```suggestion
      _.pullAll(changes, changesToRestore)
```

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  isCustomObjectSync,
+  isFieldOfTaskOrEvent,
+} from './utils'
+import { findMatchingActivityChange } from '../change_validators/task_or_event_fields_modifications'
+
+const { isDefined } = values
+const { awu } = collections.asynciterable
+
+const filterCreator: LocalFilterCreator = ({ config }) => {
+  let changesToRestore: Change[]
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:27]:
```suggestion
  let taskOrEventFieldChanges: Change[]
```

**Review by tamtamirr — COMMENTED** [review]
Amazing work 👑 Added some nits and suggestions

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts (hunk 3 above)
Comment (Nurdok) [reply packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:45]: Done

File: packages/salesforce-adapter/test/filters/utils.test.ts [hunk 7]
```diff
@@ -1405,4 +1407,28 @@ describe('filter utils', () => {
...
+    it('should return true for custom fields', () => {
+      expect(isCustomField(mockTypes.User.fields.Manager__c)).toBeTrue()
+    })
+    it('should return false for non custom fields', () => {
+      expect(isCustomField(mockTypes.AssignmentRules.fields.assignmentRule)).toBeFalse()
+    })
+  })
+
+  describe('isFieldOfTaskOrEvent', () => {
+    const activity = createCustomObjectType('event', {})
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/filters/utils.test.ts:1421]:
```suggestion
    const activity = createCustomObjectType('Activity', {})
```

File: packages/salesforce-adapter/test/filters/utils.test.ts [hunk 8]
```diff
@@ -1405,4 +1407,28 @@ describe('filter utils', () => {
       })
     })
   })
+
+  describe('isCustomField', () => {
+    it('should return true for custom fields', () => {
+      expect(isCustomField(mockTypes.User.fields.Manager__c)).toBeTrue()
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/filters/utils.test.ts:1413]:
```suggestion
      expect(mockTypes.User.fields.Manager__c).toSatisfy(isCustomField)
```
nit: Consider the following styling from jest-extended for more indicative matcher error messages

File: packages/salesforce-adapter/test/filters/utils.test.ts [hunk 9]
```diff
@@ -1405,4 +1407,28 @@ describe('filter utils', () => {
...
+    })
+    it('should return false for non custom fields', () => {
+      expect(isCustomField(mockTypes.AssignmentRules.fields.assignmentRule)).toBeFalse()
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/filters/utils.test.ts:1416]: Same comment as above, consider using `not.toSatisfy(isCustomField)`

File: packages/salesforce-adapter/test/filters/utils.test.ts [hunk 10]
```diff
@@ -1405,4 +1407,28 @@ describe('filter utils', () => {
...
+    })
+  })
+
+  describe('isFieldOfTaskOrEvent', () => {
+    const activity = createCustomObjectType('event', {})
+    const task = createCustomObjectType('Task', {})
+    const event = createCustomObjectType('Event', {})
+
+    it('should return true for fields of Task or Event', () => {
+      expect(isFieldOfTaskOrEvent(task.fields.Name)).toBeTrue()
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/filters/utils.test.ts:1426]: nit: consider using `toSatisfy` here as-well

File: packages/salesforce-adapter/test/utils.ts [hunk 11]
```diff
@@ -45,13 +45,14 @@ export const createField = (
   fieldType: TypeElement,
   fieldApiName: string,
   additionalAnnotations?: Values,
+  fieldName?: string,
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/utils.ts:48]:
```suggestion
  fieldName: string = 'field',
```

File: packages/salesforce-adapter/test/utils.ts [hunk 12]
```diff
@@ -45,13 +45,14 @@ export const createField = (
...
 ): Field => {
-  const newField = new Field(parent, 'field', fieldType, {
+  const newField = new Field(parent, fieldName ?? 'field', fieldType, {
     [constants.API_NAME]: fieldApiName,
     modifyMe: 'modifyMe',
     ...additionalAnnotations,
   })
-  parent.fields.field = newField
+  _.set(parent.fields, fieldName ?? 'field', newField)
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/utils.ts:55]:
```suggestion
  parent.fields[fieldName] = newField
```

File: packages/salesforce-adapter/test/utils.ts (hunk 11 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/test/utils.ts:48]: Nit: should simplify the code IMO

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    onFetch: ensureSafeFilterFetch({
+      warningMessage:
+        'Error occurred when attempting to remodel CustomFields of Task and Event to reference their respective Activity fields.',
+      filterName: 'taskAndEventCustomFields',
+      config,
+      fetchFilterFunc: async (elements: Element[]) => {
+        const elementsSource = buildElementsSourceForFetch(elements, config)
+        const activity = await elementsSource.get(new ElemID(SALESFORCE, ACTIVITY_CUSTOM_OBJECT))
+        if (!isCustomObjectSync(activity)) {
+          return
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:50]:
```suggestion
          log.debug('Activity Object not found')
          return
```
should help us debug fetches

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+            const getMatchingField = (objectName: string): Field | undefined =>
+              Object.values(elementsByApiName[objectName]?.fields).find(
+                field => apiNameSync(field, true) === activityFieldName,
+              )
+            return [
+              [activityField, getMatchingField(TASK_CUSTOM_OBJECT)],
+              [activityField, getMatchingField(EVENT_CUSTOM_OBJECT)],
+            ]
+          })
+          .filter((fields): fields is Field[] => fields.every(isDefined))
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:71]: What happens if either of the Objects `Task` / `Event` is excluded from the Workspace? We will lose the transformation on the fields of the latter.

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+          .forEach(([activityField, taskOrEventField]: Field[]) => {
+            const annotationsToOmit = new Set(Object.keys(activity.fields[taskOrEventField.name].annotations))
+            annotationsToOmit.delete('apiName')
+            taskOrEventField.annotations = {
+              ..._.omit(taskOrEventField.annotations, Array.from(annotationsToOmit)),
+              activityField: new ReferenceExpression(activityField.elemID),
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:77]:
```suggestion
              activityField: new ReferenceExpression(activityField.elemID, activityField),
```
We could have a resolved reference here :) and anytime we can. No reason to resolve it later if we'd want to use this reference later in the fetch process in the future.

File: packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        .filter(isFieldChange)
+        .filter(change => isFieldOfTaskOrEvent(getChangeData(change)))
+        .filter(change => isCustomField(getChangeData(change)))
+
+      _.pullAll(changes, taskOrEventFieldChanges)
+    },
+    onDeploy: async (changes: Change[]): Promise<void> => {
+      changes.push(
+        ...taskOrEventFieldChanges.filter(change => findMatchingActivityChange(change, changes) !== undefined),
+      )
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/task_and_event_custom_fields.ts:93]:
```suggestion
taskOrEventFieldChanges
    .filter(change => findMatchingActivityChange(change, changes) !== undefined)
    .forEach(change => changes.push(change))
```
We should avoid using the spread operator on arrays to avoid stack-overflows.

**Review by tamtamirr — COMMENTED** [review]
Looks great ❤️  Added some really small comments

File: packages/salesforce-adapter/jest.config.js [hunk 17]
```diff
@@ -16,7 +16,7 @@ module.exports = deepMerge(require('../../jest.base.config.js'), {
     : undefined,
   coverageThreshold: {
     global: {
-      branches: 88,
+      branches: 87,
```
Comment (tamtamirr) [inline packages/salesforce-adapter/jest.config.js:19]: Why this change?

**Review by tamtamirr — COMMENTED** [review]
LGTM! 1 Small question on the jest coverage change.

File: packages/salesforce-adapter/jest.config.js (hunk 17 above)
Comment (Nurdok) [reply packages/salesforce-adapter/jest.config.js:19]: This PR was on like 87.98% coverage. It complained about two non-covered line in tests but I had definitely covered them, so I didn't really have a way to keep it at 88.

Comment (tamtamirr) [reply packages/salesforce-adapter/jest.config.js:19]: Okay, so let's downgrade to 87.9, 1% is a drastic decrease :)

**Review by tamtamirr — APPROVED** [review]
LGTM Great Work 👑
